src/RwaUrn.sol

```diff
@@ -0,0 +1,125 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.5.12;
+
+import "lib/dss-interfaces/src/dss/VatAbstract.sol";
+import "lib/dss-interfaces/src/dapp/DSTokenAbstract.sol";
+import "lib/dss-interfaces/src/dss/GemJoinAbstract.sol";
+import "lib/dss-interfaces/src/dss/DaiJoinAbstract.sol";
+import "lib/dss-interfaces/src/dss/DaiAbstract.sol";
+
+contract RwaUrn {
+    // --- auth ---
+    mapping (address => uint256) public wards;
+    mapping (address => uint256) public can;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "RwaUrn/not-authorized");
+        _;
+    }
+    function hope(address usr) external auth {
+        can[usr] = 1;
+        emit Hope(usr);
+    }
+    function nope(address usr) external auth {
+        can[usr] = 0;
+        emit Nope(usr);
+    }
+    modifier operator {
+        require(can[msg.sender] == 1, "RwaUrn/not-operator");
+        _;
+    }
+
+    VatAbstract  public vat;
+    GemJoinAbstract public gemJoin;
+    DaiJoinAbstract public daiJoin;
+    address public outputConduit;
+
+    // Events
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Hope(address indexed usr);
+    event Nope(address indexed usr);
+    event File(bytes32 indexed what, address data);
+    event Lock(uint256 wad);
+    event Free(uint256 wad);
+    event Draw(uint256 wad);
+    event Wipe(uint256 wad);
+
+    // --- init ---
+    constructor(
+        address vat_, address gemJoin_, address daiJoin_, address outputConduit_
+    ) public {
+        // requires in urn that outputConduit isn't address(0)
+        vat = VatAbstract(vat_);
+        gemJoin = GemJoinAbstract(gemJoin_);
+        daiJoin = DaiJoinAbstract(daiJoin_);
+        outputConduit = outputConduit_;
+        wards[msg.sender] = 1;
+        DSTokenAbstract(gemJoin.gem()).approve(address(gemJoin), uint256(-1));
+        DaiAbstract(daiJoin.dai()).approve(address(daiJoin), uint256(-1));
+        VatAbstract(vat_).hope(address(daiJoin));
+        emit Rely(msg.sender);
+    }
+
+    // --- administration ---
+    function file(bytes32 what, address data) external auth {
+        if (what == "outputConduit") { outputConduit = data; }
+        else revert("RwaUrn/unrecognised-param");
+        emit File(what, data);
+    }
+
+    // --- cdp operation ---
+    // n.b. that the operator must bring the gem
+    function lock(uint256 wad) external operator {
+        require(wad <= 2**255 - 1, "RwaUrn/overflow");
+        DSTokenAbstract(gemJoin.gem()).transferFrom(msg.sender, address(this), wad);
```

Comment:
Wrap this in a `require()` for extra safety.

Reply:
Won't do since governance deploy's the token and can make guarantees here.